spring-cloud-data-module-deployers/spring-cloud-data-module-deployer-spi/src/main/java/org/springframework/cloud/data/module/ModuleKey.java

```diff
@@ -0,0 +1,161 @@
+ * Methods {@link #toUID()} and {@link #fromUID(String)} can be used to convert
+ * a key to a string and a string to a key, respectively. The UID string may be
+ * used to uniquely identify a module in a database or execution environment.
+ *
+ * @author Patrick Peralta
+ */
+public class ModuleKey implements Serializable {
+	private static final long serialVersionUID = 1L;
+
+	/**
+	 * The name of the associated group or deployment unit this
+	 * module belongs to, such as a stream or job.
+	 */
+	private final String group;
+
+	/**
+	 * The name of the module as defined in the module registry,
+	 * such as {@code http} or {@code file}.
+	 */
+	private final String name;
+
+	/**
+	 * The label provided to uniquely identify the module as part
+	 * of group membership. For instance, if a group has multiple
+	 * {@code filter} modules, each will have a unique label.
+	 */
+	private final String label;
+
+	/**
+	 * Construct a {@code ModuleKey}.
+	 *
+	 * @param group name of group (stream/job) this module belongs to
+	 * @param name  name of module as defined in module registry
+	 * @param label label to uniquely identify this module in its group
+	 */
+	public ModuleKey(String group, String name, String label) {
+		Assert.hasText(group);
+		Assert.hasText(name);
+		Assert.hasText(label);
+		Assert.doesNotContain(group, ".");
+		Assert.doesNotContain(name, ".");
+		Assert.doesNotContain(label, ".");
+		this.group = group;
+		this.name = name;
+		this.label = label;
+	}
+
+	/**
+	 * @see #group
+	 */
+	public String getGroup() {
+		return group;
+	}
+
+	/**
+	 * @see #name
+	 */
+	public String getName() {
+		return name;
+	}
+
+	/**
+	 * @see #label
+	 */
+	public String getLabel() {
+		return label;
+	}
+
+	@Override
+	public boolean equals(Object o) {
+		if (this == o) {
+			return true;
+		}
+		if (o == null || getClass() != o.getClass()) {
+			return false;
+		}
+
+		ModuleKey that = (ModuleKey) o;
+		return this.group.equals(that.group)
+				&& this.name.equals(that.name)
+				&& this.label.equals(that.label);
+	}
+
+	@Override
+	public int hashCode() {
+		int result = group.hashCode();
+		result = 31 * result + name.hashCode();
+		result = 31 * result + label.hashCode();
+		return result;
+	}
+
+	/**
+	 * Return a string containing the key fields separated by
+	 * periods. This string may be used as a key in a database
+	 * to uniquely identify a module.
+	 *
+	 * @return string representation of this key
+	 */
+	public String toUID() {
+		return String.format("%s.%s.%s", this.group, this.name, this.label);
+	}
+
+	/**
+	 * Parse the given string and return a {@code ModuleKey} based
+	 * on the string contents.
+	 *
+	 * @param uid uid containing the fields for a {@code ModuleKey}
+	 * @return a new {@code ModuleKey} based on the provided string
+	 * @throws NullPointerException if a null uid is provided
+	 * @throws IllegalArgumentException if an invalid uid is provided
+	 */
+	public static ModuleKey fromUID(String uid) {
```

Comment:
I wonder if this might get confusing if users expect UUID format. Perhaps this method could just be called parse? The corresponding toUID() could just be this class' implementation of toString().